pkg/controllers/apps/feedinventory/feedinventory.go

```diff
@@ -0,0 +1,366 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package feedinventory
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	utilpointer "k8s.io/utils/pointer"
+
+	appsapi "github.com/clusternet/clusternet/pkg/apis/apps/v1alpha1"
+	clusternetclientset "github.com/clusternet/clusternet/pkg/generated/clientset/versioned"
+	appinformers "github.com/clusternet/clusternet/pkg/generated/informers/externalversions/apps/v1alpha1"
+	applisters "github.com/clusternet/clusternet/pkg/generated/listers/apps/v1alpha1"
+	"github.com/clusternet/clusternet/pkg/known"
+	"github.com/clusternet/clusternet/pkg/utils"
+)
+
+var subKind = appsapi.SchemeGroupVersion.WithKind("Subscription")
+
+// Controller is a controller that handle HelmRelease
+type Controller struct {
+	clusternetClient clusternetclientset.Interface
+
+	// workqueue is a rate limited work queue. This is used to queue work to be
+	// processed instead of performing it as soon as a change happens. This
+	// means we can ensure we only process a fixed amount of resources at a
+	// time, and makes it easy to ensure we are never processing the same item
+	// simultaneously in two different workers.
+	workqueue workqueue.RateLimitingInterface
+
+	subLister      applisters.SubscriptionLister
+	subSynced      cache.InformerSynced
+	manifestLister applisters.ManifestLister
+	manifestSynced cache.InformerSynced
+
+	recorder record.EventRecorder
+	registry Registry
+
+	// namespace where Manifests are created
+	reservedNamespace string
+}
+
+func NewController(clusternetClient clusternetclientset.Interface,
+	subsInformer appinformers.SubscriptionInformer, manifestInformer appinformers.ManifestInformer,
+	recorder record.EventRecorder, registry Registry, reservedNamespace string) (*Controller, error) {
+	c := &Controller{
+		clusternetClient:  clusternetClient,
+		workqueue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "feedInventory"),
+		subLister:         subsInformer.Lister(),
+		subSynced:         subsInformer.Informer().HasSynced,
+		manifestLister:    manifestInformer.Lister(),
+		manifestSynced:    manifestInformer.Informer().HasSynced,
+		recorder:          recorder,
+		registry:          registry,
+		reservedNamespace: reservedNamespace,
+	}
+
+	// For FeedInventory, we only need to care about the changes of Subscription
```

Comment:
What if a workload has been scaled up but the subscription has not been modified? It seems that this controller does not care about the events of original workload. E.g., a deployment with 5 desired replicas is now been scaled up to 10 desired replicas, while the corresponding subscription is as same as before.

Reply:
Make sense. Addressed.